// npm-increaser-downloads/lib/src/cli.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const yargs_1 = __importDefault(require("yargs"));
const helpers_1 = require("yargs/helpers");
const prompts_1 = require("./cli/prompts");
const config_1 = require("./config");
const spammer_1 = require("./spammer/spammer");
const argv = (0, yargs_1.default)((0, helpers_1.hideBin)(process.argv))
    .usage("Usage: nid [options]\n\n" +
    "Note: if no options are provided, nid will prompt for input interactively.\n\n" +
    "For advanced configuration, use npm-increaser-downloads.config.ts in the root directory.\n" +
    "See also the sample: https://github.com/MinhOmega/npm-increaser-downloads#readme\n\n" +
    "Examples:\n\n" +
    "  $ nid\n" +
    "  $ nid -p my-package -n 1000 -m 300 -t 3000\n" +
    "  $ nid --package-name my-package --num-downloads 1000 --max-concurrent-downloads 300 --download-timeout 3000\n\n" +
    "All options are documented under: nid help")
    .option("package-name", {
    alias: "p",
    type: "string",
    description: "NPM package to increase the downloads of",
})
    .option("num-downloads", {
    alias: "n",
    type: "number",
    description: "Number of times to download the package",
})
    .option("max-concurrent-downloads", {
    alias: "m",
    type: "number",
    description: "Amount of downloads to run in parallel at once",
})
    .option("download-timeout", {
    alias: "t",
    type: "number",
    description: "Max time (in ms) to wait for a download to complete",
})
    .help()
    .alias("help", "h")
    .epilogue("For more information, visit https://github.com/MinhOmega/npm-increaser-downloads").argv;
const runWithArgs = (args) => {
    // Check if any relevant argument was provided
    const hasAnyArg = ["package-name", "num-downloads", "max-concurrent-downloads", "download-timeout"].some((arg) => args[arg] !== undefined);
    if (!hasAnyArg) {
        // If no relevant arguments were provided, fall back to CLI prompts
        (0, prompts_1.getConfigFromCli)().then(config_1.setConfig).then(spammer_1.run);
        return;
    }
    const config = {
        packageName: args["package-name"],
        numDownloads: args["num-downloads"],
        maxConcurrentDownloads: args["max-concurrent-downloads"],
        downloadTimeout: args["download-timeout"],
    };
    const isConfigComplete = Object.values(config).every((value) => value !== undefined);
    if (isConfigComplete) {
        (0, config_1.setConfig)(config);
        (0, spammer_1.run)();
    }
    else {
        // If any argument was provided but config is incomplete, show help and exit
        console.log("Incomplete configuration. Please provide all required options or run without arguments for interactive mode.");
        yargs_1.default.showHelp();
        process.exit(1);
    }
};
runWithArgs(argv);
