// npm-increaser-downloads/lib/src/models/stats.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stats = void 0;
const config_1 = require("../config");
class Stats {
    constructor(startTime, successfulDownloads, failedDownloads) {
        this.successfulDownloads = 0;
        this.failedDownloads = 0;
        this.startTime = startTime;
        this.successfulDownloads = successfulDownloads ?? this.successfulDownloads;
        this.failedDownloads = failedDownloads ?? this.failedDownloads;
    }
    getDownloadSpeed() {
        const currentTime = Date.now();
        const timeElapsed = currentTime - this.startTime;
        const downloadsPerMillisecond = this.successfulDownloads / timeElapsed;
        const downloadsPerSecond = downloadsPerMillisecond * 1000;
        if (downloadsPerSecond > 0 && isFinite(downloadsPerSecond)) {
            return downloadsPerSecond;
        }
        else {
            return 0;
        }
    }
    getTimeRemaining() {
        const downloadsRemaining = (0, config_1.getConfig)().numDownloads - this.successfulDownloads;
        const downloadsPerSecond = this.getDownloadSpeed();
        if (downloadsPerSecond <= 0) {
            return null;
        }
        else {
            const secondsRemaining = downloadsRemaining / downloadsPerSecond;
            return secondsRemaining;
        }
    }
}
exports.Stats = Stats;
