// npm-increaser-downloads/lib/src/cli/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logError = exports.logComplete = exports.logDownload = exports.mapToDate = exports.mapToString = exports.terminalSpinner = void 0;
const date_fns_1 = require("date-fns");
const ora_1 = __importDefault(require("ora"));
const config_1 = require("../config");
exports.terminalSpinner = (0, ora_1.default)();
const mapToString = (num) => {
    if (Number.isInteger(num)) {
        return num.toString();
    }
    else {
        return num.toFixed(2);
    }
};
exports.mapToString = mapToString;
const mapToDate = (seconds) => {
    if (seconds === null || seconds < 0) {
        return "--:--:--";
    }
    const milliseconds = 1000 * seconds;
    const date = new Date(milliseconds);
    const hours = (0, date_fns_1.secondsToHours)(seconds);
    const hoursString = hours.toString().padStart(2, "0");
    const minutesSecondsString = (0, date_fns_1.format)(date, "mm:ss");
    return `${hoursString}:${minutesSecondsString}`;
};
exports.mapToDate = mapToDate;
const logDownload = (stats) => {
    if (!exports.terminalSpinner.isSpinning) {
        exports.terminalSpinner.start();
    }
    exports.terminalSpinner.text = `\n`;
    exports.terminalSpinner.text += `Download count:            ${(0, exports.mapToString)(stats.successfulDownloads)}/${(0, config_1.getConfig)().numDownloads}\n`;
    exports.terminalSpinner.text += `Download speed:            ${(0, exports.mapToString)(stats.getDownloadSpeed())} dl/s\n`;
    exports.terminalSpinner.text += `Estimated time remaining:  ${(0, exports.mapToDate)(stats.getTimeRemaining())}\n`;
};
exports.logDownload = logDownload;
const logComplete = () => {
    exports.terminalSpinner.succeed(`Completed ${(0, config_1.getConfig)().numDownloads} downloads for ${(0, config_1.getConfig)().packageName}`);
};
exports.logComplete = logComplete;
const logError = (error) => {
    exports.terminalSpinner.stopAndPersist();
    exports.terminalSpinner.fail(error.toString());
};
exports.logError = logError;
