// npm-increaser-downloads/lib/src/cli/prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigFromCli = exports.getDownloadTimeout = exports.getMaxConcurrentDownloads = exports.getNumberOfDownloads = exports.getPackageName = void 0;
const prompts_1 = require("@inquirer/prompts");
const validators_1 = require("./validators");
const getPackageName = async () => {
    return (0, prompts_1.input)({
        message: "Package name: ",
        validate: (name) => (0, validators_1.validatePackageName)(name),
    });
};
exports.getPackageName = getPackageName;
const getNumberOfDownloads = async () => {
    const result = await (0, prompts_1.number)({
        message: "Number of downloads: ",
        default: 1000,
        validate: (downloads) => (0, validators_1.validateNumbers)(downloads),
    });
    return result ?? 1000;
};
exports.getNumberOfDownloads = getNumberOfDownloads;
const getMaxConcurrentDownloads = async () => {
    const result = await (0, prompts_1.number)({
        message: "Number of concurrent downloads: ",
        default: 300,
        validate: (downloads) => (0, validators_1.validateNumbers)(downloads),
    });
    return result ?? 300;
};
exports.getMaxConcurrentDownloads = getMaxConcurrentDownloads;
const getDownloadTimeout = async () => {
    const result = await (0, prompts_1.number)({
        message: "Time to wait for a download to complete (in ms): ",
        default: 3000,
        validate: (downloads) => (0, validators_1.validateNumbers)(downloads),
    });
    return result ?? 3000;
};
exports.getDownloadTimeout = getDownloadTimeout;
const getEmptyConfig = () => {
    return {
        packageName: "",
        numDownloads: 0,
        maxConcurrentDownloads: 0,
        downloadTimeout: 0,
    };
};
const getConfigFromCli = async () => {
    const config = getEmptyConfig();
    config.packageName = await (0, exports.getPackageName)();
    config.numDownloads = await (0, exports.getNumberOfDownloads)();
    config.maxConcurrentDownloads = await (0, exports.getMaxConcurrentDownloads)();
    config.downloadTimeout = await (0, exports.getDownloadTimeout)();
    return config;
};
exports.getConfigFromCli = getConfigFromCli;
