// npm-increaser-downloads/lib/src/spammer/spammer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.downloadPackage = exports.getVersionPackage = exports.getNpmjsResponse = exports.getNpmsioResponse = void 0;
const gaxios_1 = require("gaxios");
const logger_1 = require("../cli/logger");
const config_1 = require("../config");
const stats_model_1 = require("../models/stats.model");
const utils_1 = require("./utils");
const logger_2 = require("../cli/logger");
const getNpmsioResponse = async () => {
    const encodedPackageName = (0, utils_1.getEncodedPackageName)((0, config_1.getConfig)().packageName);
    const npmsioResponse = await (0, gaxios_1.request)({
        baseUrl: "https://api.npms.io",
        url: `/v2/package/${encodedPackageName}`,
        method: "GET",
    }).catch((response) => {
        throw Error(`Failed to download ${response.config.url}\n${response.message}`);
    });
    return npmsioResponse.data;
};
exports.getNpmsioResponse = getNpmsioResponse;
const getNpmjsResponse = async () => {
    const encodedPackageName = (0, utils_1.getEncodedPackageName)((0, config_1.getConfig)().packageName);
    const npmsResponse = await (0, gaxios_1.request)({
        baseUrl: "https://registry.npmjs.org",
        url: `/${encodedPackageName}/latest`,
        method: "GET",
    }).catch((response) => {
        throw Error(`Failed to download ${response.config.url}\n${response.message}`);
    });
    return npmsResponse.data;
};
exports.getNpmjsResponse = getNpmjsResponse;
const getVersionPackage = async () => {
    try {
        logger_2.terminalSpinner.start();
        const npmioResponse = await (0, exports.getNpmsioResponse)();
        logger_2.terminalSpinner.succeed(`Package version found on npms.io with version ${npmioResponse.collected.metadata.version}`);
        return npmioResponse.collected.metadata.version;
    }
    catch (npmioError) {
        logger_2.terminalSpinner.text = "Package not found in npms.io, trying npmjs.com...";
        try {
            const npmjsResponse = await (0, exports.getNpmjsResponse)();
            logger_2.terminalSpinner.succeed(`Package found on npmjs.com with version ${npmjsResponse.version}`);
            return npmjsResponse.version;
        }
        catch (npmjsError) {
            logger_2.terminalSpinner.fail("Failed to get package version");
            throw new Error(`Failed to get package version: ${npmioError.message}, ${npmjsError.message}`);
        }
    }
};
exports.getVersionPackage = getVersionPackage;
const downloadPackage = async (version, stats) => {
    const packageName = (0, config_1.getConfig)().packageName;
    const unscopedPackageName = (0, utils_1.stripOrganisationFromPackageName)(packageName);
    return (0, gaxios_1.request)({
        baseUrl: "https://registry.yarnpkg.com",
        url: `/${packageName}/-/${unscopedPackageName}-${version}.tgz`,
        method: "GET",
        timeout: (0, config_1.getConfig)().downloadTimeout,
        responseType: "stream",
    })
        .then(() => {
        stats.successfulDownloads++;
    })
        .catch((error) => {
        stats.failedDownloads++;
        console.error("Download failed:", error.message);
    });
};
exports.downloadPackage = downloadPackage;
const spamDownloads = async (version, stats) => {
    const batchSize = Math.min((0, config_1.getConfig)().maxConcurrentDownloads, 50); // Limit to 50 concurrent downloads
    const totalDownloads = (0, config_1.getConfig)().numDownloads;
    for (let i = 0; i < totalDownloads; i += batchSize) {
        const batch = Math.min(batchSize, totalDownloads - i);
        const requests = [];
        for (let j = 0; j < batch; j++) {
            requests.push((0, exports.downloadPackage)(version, stats));
        }
        await Promise.all(requests);
        // Add a small delay between batches to allow for connection cleanup
        await new Promise((resolve) => setTimeout(resolve, 100));
    }
};
/**
 * Runs the download spammer.
 * @returns A Promise that resolves when the spamming is complete.
 */
const run = async () => {
    try {
        const version = await (0, exports.getVersionPackage)();
        const startTime = Date.now();
        const stats = new stats_model_1.Stats(startTime);
        const loggingInterval = setInterval(() => (0, logger_1.logDownload)(stats), 1000);
        await spamDownloads(version, stats);
        clearInterval(loggingInterval);
        (0, logger_1.logComplete)();
    }
    catch (e) {
        (0, logger_1.logError)(e);
    }
};
exports.run = run;
